### ISSUE TITLE
End of the session

### ISSUE THREAD

**Issue by hash55** [theOehrly/Fast-F1#43]
Hi! Is there any method to know when does a session end?
Especially for qualifying sessions, just wanted to know at what time of the session does the chequered flag appeared.
(it is to know if a driver crossed the line on time and by how much)
Thanks

**Comment by theOehrly** [comment]
Yes, there is. There is an api method for that in `fastf1.api`. See the docs here: https://theoehrly.github.io/Fast-F1/api.html#fastf1.api.session_status_data

You should be able to get the information you want from the data that it returns.
The function requires a parameter `api_path`. You can get that from a `Session` object. Therefore, you first create a session by calling `get_session` and then you can call `fastf1.api.session_status_data(session.api_path)`.

Is that what you are looking for?

**Comment by hash55** [comment]
thanks for the reply! That's exactly what I was looking for. By the way, how accurate is this data?(That microseconds are right?)

**Comment by theOehrly** [comment]
Good question, but I'm sorry to say that I don't know for sure. I can just make assumptions there.
In short, I don't think the milliseconds are correct. 

If you check start and end of a qualifying round, then you can see that the session would sometimes be close to half a tenth of a second shorter than the official length.

Now, the actual timing system at the track needs to be very capable, and I'm fairly sure that a session is exactly as long as it should be. 
I do think that those timestamps denote the time at which their (way less critical) server system processed this event. I do know for sure that the same timestamps for laps are not accurate (lap start and end time). After all, it is not actually relevant, the way that they're presenting this data in the app.

I do optimize the laps timestamps as much as possible. But I can only minimize random jitter. There might be a constant offset by which all laps timestamps are off. I don't know, because I haven't yet found a good reference to compare that against.

And just to make this clear, the length of a lap, i.e. the lap time and sector times are 100% accurate. Those are the only data that I fully trust and I try to use it as much as possible to correct the timestamps for example.

If you have the time and motivation, you should try to find some reference and evaluate how accurate this is. I'd actually be very interested in the result.

**Comment by hash55** [comment]
Thanks again. Of course I would try to dig into it. But I don't know why the t0_date function always returns "none"
`print(fastf1.get_session('2021','Spanish Grand Prix','Q').t0_date)`
Am I doing something wrong?

**Comment by theOehrly** [comment]
No, not your fault. That's just badly documented, actually. You need to load laps **with telemetry** so that the t0_date can be calculated. The telemetry data is required for that, as inconvenient as it is.

**Comment by hash55** [comment]
Ok thanks. The problem of that is that telemetry is just available from 2018, isn't it? I was trying to search more in historical races and those things

**Comment by theOehrly** [comment]
Yes, correct. Unfortunately, there is no data earlier than the 2018 season

**Comment by theOehrly** [comment]
What I'm guessing is that the timestamps provided by the api are not created by the official timing system. I suppose they maybe are created when their api server processes the event. And there is no focus put on having them be fully accurate.

I do wish that all timestamps were accurate to the millisecond and that all data is perfectly synchronized. But it is just not possible. For lap start and end times a lot of effort is made to correct inaccuracies.
The `session_status_data` and `track_status_data` timestamps are not modified though. They are used exactly as they are received from the api. And I don't really see a way to make them more accurate.

I do find it curious that a race is always 220ms too long. 220ms is also the sample rate of the position data. Maybe this is just the interval at which their server processes events. And the end of the session is somehow processed one cycle too late. 
Have you summed up the lap times of one drive for comparison? Does that match what they display on TV as the duration of the race?

**Comment by theOehrly** [comment]
@hash55 sorry for responding incredibly late. I needed to actually take some time to write a proper answer here. Anyway, synchronizing the data is a very interesting topic. But it is made difficult by our lack of data, accuracy and knowledge about the data.


First, do you know how close Gasly was (approximately) to not starting the lap in time? How far off is the timing of FastF1?


Second, I have previously tried various approaches to synchronize the telemetry data. 
One approach was to use the finish line speed trap as a reference point. While this did work in principle, it did not work better  (see: https://www.f1technical.net/forum/viewtopic.php?p=919051#p919051).


The other method I tried was to synchronize the lap time to the finish line position (and sector position). The problem here is that we don't know the position of the finish line. The only assumption that can be made is, that the position stays the same.
What I did was to introduce a time offset at the beginning of all laps as a variable. The coordinates of the finish line crossing should be the same again every lap when adding the subsequent lap times. In theory, there should be one specific correct offset for which the finish line is in exactly the same place every lap. But you would expect the position to vary slightly for all other offsets. Therefore, this is a minimization problem with one dependent variable. 
This approach did work in principle, too. But there was no one clear solution. The result was a range of coordinates for which the position jitter was close to zero, but not fully zero. That range was too wide to make any useful conclusions.

I did put this down to various issues. Among others, the low sample rate of the data. I also assumed that the car's internal clocks are not synchronized. This has now been confirmed by the AWS blog post that you have linked to.

With this additional uncertainty, there are now ten additional dependent variables to solve for, assuming that teams synchronize the clocks between their own cars. In theory, we have way more data than required to solve this problem. But the low sample rate (and potentially other things) introduce uncertainty to the data. Even in the AWS post, they describe how they calculate the reaction time of a driver as the median of a distribution of results.
I don't want to say that it is not possible, but exploring how feasible and accurate this is requires a lot of time.


Also, there may be other possibilities for synchronizing the data. 


As an additional problem, we don't know if the timestamps we get are actually the ones created by the cars' on-board computer. Maybe they're generated by the API server when processing the data? 


With regard to all these problems, there is currently no active synchronization of the data done. Only the lap start and end times are adjusted to match the lap times. But if we could actually synchronize the data properly, this would be much better.
